datastore/postgres/shop_repository.go

```diff
@@ -11,19 +13,82 @@ type shopRepository struct {
 	executor sql_datastore.Executor
 }
 
-var shops = []domain.Shop{
-	domain.Shop{
-		Name: "unverpackt",
-	},
-	domain.Shop{
-		Name: "unverpackt2",
-	},
-}
-
 func NewShopRepository(executor sql_datastore.Executor) *shopRepository {
 	return &shopRepository{executor: executor}
 }
 
 func (s *shopRepository) FindShopsByQuery(ctx context.Context, search string, lat float32, long float32) ([]domain.Shop, error) {
-	return shops, nil
+	var shops []domain.Shop
+	var productIds []string
+
+	sql, args, err := queryBuilder.
+		Select("products.id").
+		Distinct().
+		From("products").
+		Join("categories_translations on categories_translations.category_id = products.category_id").
+		Join("products_translations on products_translations.product_id = products.id").
+		Where(sq.Or{
+			sq.ILike{"categories_translations.value": "%" + search + "%"},
+			sq.ILike{"products_translations.value": "%" + search + "%"},
+		}).ToSql()
+	if err != nil {
+		log.Printf("creating sql statement for matching products failed: %v", err)
+		return nil, err
+	}
+
+	rows, err := s.executor.QueryxContext(ctx, sql, args...)
+
+	if err != nil {
+		log.Printf("error when executing query for finding matching product ids", err)
+		return nil, err
+	}
+
+	defer rows.Close()
+
+	for rows.Next() {
+		var id string
+		err = rows.Scan(&id)
+		if err != nil {
+			return nil, err
+		}
+
+		productIds = append(productIds, id)
+	}
+
+	if len(productIds) == 0 {
+		return nil, nil
+	}
+
+	sql, args, err = queryBuilder.
+		Select("shops.id", "shops.name", "shops.address", "shops.lat", "shops.long").
+		Distinct().
+		From("shops").
+		Join("shops_products on shops.id = shops_products.shop_id").
+		Where(sq.Eq{"shops_products.product_id": productIds}).ToSql()
+
+	if err != nil {
+		log.Printf("creating sql statement for finding shops by query failed: %v", err)
+		return nil, err
+	}
+
+	rows, err = s.executor.QueryxContext(ctx, sql, args...)
+
+	if err != nil {
+		log.Println("error when executing query for finding shops by query", err)
+		return nil, err
+	}
+
+	defer rows.Close()
+
+	for rows.Next() {
+		var shop Shop
+		err = rows.StructScan(&shop)
+		if err != nil {
+			return shops, err
+		}
+
+		shops = append(shops, *shop.toDomain())
+	}
+
+	return shops, err
```

Comment:
In the best case we can have one combined query here:

```sql
select
  distinct shops.id,
  shops.name,
  shops.address,
  shops.lat,
  shops.long
from shops
join shops_products on shops.id = shops_products.shop_id
where
  shops_products.product_id in (
    select
      distinct products.id
    from products
    join categories_translations on categories_translations.category_id = products.category_id
    join products_translations on products_translations.product_id = products.id
    where
      categories_translations.value ilike '%$1%'
      or products_translations.value ilike '%$1%'
  )
```

Reply:
Yura would probably have more expertise here, so I'd defer to him.